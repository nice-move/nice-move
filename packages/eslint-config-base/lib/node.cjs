'use strict';

const { configHas } = require('./utils.cjs');

const commonjs =
  configHas(
    ({ commonjs: globs = [] }) => globs,
    (globs) => globs,
  ) || [];

module.exports = {
  plugins: ['n'],
  rules: {
    'n/hashbang': 'error',
    'n/no-process-env': 'error',
    'n/no-path-concat': 'error',
  },
  overrides: [
    {
      files: '**/*',
      excludedFiles: ['*.cjs', '*.cts', ...commonjs],
      rules: {
        'import/no-commonjs': 'error',
      },
    },
    {
      files: '*.nb.*',
      env: {
        browser: false,
        node: true,
        commonjs: true,
      },
      rules: {
        'no-unused-vars': 'warn',
        'import/no-extraneous-dependencies': 'off',
        'import/no-commonjs': 'off',
        'import/no-nodejs-modules': 'off',
        'unicorn/prefer-module': 'off',
      },
    },
    {
      files: '*.{m,c}{t,j}s',
      excludedFiles: '*.nb.*',
      env: {
        node: true,
      },
      rules: {
        'n/no-deprecated-api': 'error',
        'n/no-process-env': 'off',
        'n/no-process-exit': 'warn',
        'n/no-unsupported-features/es-builtins': 'error',
        'n/no-unsupported-features/es-syntax': 'error',
        'n/no-unsupported-features/node-builtins': 'error',
        'n/prefer-global/buffer': 'warn',
        'n/prefer-global/console': 'warn',
        'n/prefer-global/process': 'warn',
        'n/prefer-global/url-search-params': 'warn',
        'n/prefer-global/url': 'warn',
        'n/process-exit-as-throw': 'error',
      },
    },
    {
      files: '*.*',
      excludedFiles: [
        '*.{m,c}ts',
        '*.cjs',
        '*.md',
        '*.nb.*',
        '*.ts',
        '**/*.md/*',
        ...commonjs,
      ],
      rules: {
        'n/file-extension-in-import': 'error',
      },
    },
    {
      files: ['*.mjs', '*.js', '*.jsx'],
      parser: '@babel/eslint-parser',
      parserOptions: {
        requireConfigFile: false,
        babelOptions: {
          babelrc: false,
          configFile: false,
          plugins: [require.resolve('@babel/plugin-syntax-import-attributes')],
        },
      },
    },
    {
      // for node.js
      files: ['*.mjs', '*.mts'],
      excludedFiles: '*.nb.*',
      env: {
        browser: false,
        commonjs: false,
      },
      globals: {
        __dirname: 'off',
        __filename: 'off',
        exports: 'off',
        module: 'off',
        require: 'off',
      },
    },
    {
      files: ['*.cjs', ...commonjs],
      excludedFiles: '*.nb.*',
      env: {
        browser: false,
        commonjs: true,
      },
      parserOptions: {
        ecmaFeatures: {
          impliedStrict: false,
        },
        sourceType: 'script',
      },
      rules: {
        strict: ['error', 'global'],
        'n/no-exports-assign': 'error',
        'n/no-new-require': 'error',
      },
    },
    {
      files: ['*.cjs', '*.cts'],
      rules: {
        'unicorn/prefer-module': 'off',
      },
    },
  ],
};
