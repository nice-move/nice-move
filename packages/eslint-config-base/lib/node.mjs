import nodePlugin from 'eslint-plugin-n';

function pick({ rules, plugins }) {
  return { rules, plugins };
}

export default [
  {
    files: ['**/*.{mjs,mts,cts}'],
    ...pick(nodePlugin.configs['flat/recommended-module']),
  },
  {
    files: ['**/*.{mjs,mts,cts}'],
    rules: {
      'n/no-missing-import': 0,
      'n/no-extraneous-import': 0,
      'n/no-missing-require': 0,
      'n/no-extraneous-require': 0,
    },
  },
  {
    files: ['**/*.cjs'],
    ...pick(nodePlugin.configs['flat/recommended-script']),
  },
  {
    files: ['**/*.cjs'],
    rules: {
      'n/no-missing-import': 0,
      'n/no-extraneous-import': 0,
    },
  },
  {
    files: ['**/*.{js,jsx,ts,tsx,mjs,mts,vue}'],
    rules: {
      'import/no-commonjs': 'error',
    },
  },
  {
    files: ['**/*.{m,c}{t,j}s'],
    rules: {
      'n/no-path-concat': 'error',
      'n/no-deprecated-api': 'error',
      'n/no-process-env': 'off',
      'n/no-process-exit': 'warn',
      'n/no-unsupported-features/es-builtins': 'error',
      'n/no-unsupported-features/es-syntax': 'error',
      'n/no-unsupported-features/node-builtins': 'error',
      'n/prefer-global/buffer': 'warn',
      'n/prefer-global/console': 'warn',
      'n/prefer-global/process': 'warn',
      'n/prefer-global/url-search-params': 'warn',
      'n/prefer-global/url': 'warn',
      'n/process-exit-as-throw': 'error',
    },
  },
  {
    files: ['**/*.{js,jsx,ts,tsx,cts,mjs,mts,vue}'],
    ignores: ['**/*.md/*'],
    rules: {
      'n/file-extension-in-import': 'error',
    },
    plugins: {
      n: nodePlugin,
    },
    settings: {
      n: {
        tryExtensions: [
          '.js',
          '.ts',
          '.jsx',
          '.tsx',
          '.mjs',
          '.mts',
          '.cjs',
          '.cts',
          '.vue',
        ],
        typescriptExtensionMap: [],
      },
    },
  },
  {
    files: ['**/*.{cjs,cts}'],
    rules: {
      'n/no-exports-assign': 'error',
      'n/no-new-require': 'error',
      'unicorn/prefer-module': 'off',
    },
  },
];
