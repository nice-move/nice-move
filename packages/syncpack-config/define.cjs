'use strict';

const { fileURLToPath } = require('node:url');
const { join } = require('node:path');
const { findWorkspaces } = require('find-workspaces');

module.exports = function defineConfig(url, config = {}) {
  let pkg = {};

  try {
    pkg = require(
      url === '~'
        ? join(process.cwd(), 'package.json')
        : join(fileURLToPath(url), '../package.json'),
    );
  } catch (error) {
    console.error(error);
  }

  const ws = pkg.packageManager?.startsWith?.('pnpm')
    ? findWorkspaces() ?? []
    : [];

  return {
    ...config,
    lintFormatting: false,
    customTypes: {
      ...config.customTypes,
      engines: {
        path: 'engines',
        strategy: 'versionsByName',
      },
      packageManager: {
        path: 'packageManager',
        strategy: 'name@version',
      },
    },
    semverGroups: [
      ...(config.semverGroups || []),
      {
        dependencies: [
          'typescript',
          'react',
          'react-dom',
          '@types/react-dom',
          '@types/react',
        ],
        range: '~',
      },
      {
        dependencyTypes: ['local', 'packageManager'],
        range: '',
      },
      {
        dependencyTypes: ['!engines'],
        range: '^',
      },
    ],
    versionGroups: [
      {
        dependencies: ['@types/**'],
        dependencyTypes: ['!peer'],
        isBanned: true,
        label: '@types packages should only be under peerDependencies',
      },
      {
        dependencies: [
          'react',
          'react-dom',
          '@types/react-dom',
          '@types/react',
        ],
        dependencyTypes: ['!local'],
        label: 'Pin react',
        policy: 'sameRange',
      },
      ws.length > 0
        ? {
            dependencies: ws.map((item) => item.package.name),
            dependencyTypes: ['!local'],
            label: 'Pin pnpm workspace',
            pinVersion: 'workspace:~',
          }
        : undefined,
      ...(config.versionGroups || []),
      {
        dependencies: ['node'],
        dependencyTypes: ['engines'],
        label: 'Pin engines.node',
        pinVersion: pkg.engines?.node || '^20.0.0',
      },
      {
        dependencyTypes: ['!local'],
        label: 'Pin others',
        preferVersion: 'highestSemver',
      },
    ].filter(Boolean),
  };
};
