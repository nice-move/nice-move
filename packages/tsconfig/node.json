{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "./base.json",
  "compilerOptions": {
    "types": [],
    "typeRoots": [
      "types",
      "node_modules/@types",
      "node_modules/@types/node"
    ]
  },
  "display": "Node"
}
