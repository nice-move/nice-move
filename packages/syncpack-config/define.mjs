/* eslint-disable unicorn/no-useless-undefined */
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { findWorkspaces } from 'find-workspaces';
import { parse } from 'yaml';

/**
 * 从 pnpm-workspace.yaml 文件中读取目录配置
 * @returns {{ catalog: object, catalogs: object }} 包含目录配置的对象
 */
function readYaml() {
  try {
    const path = join(process.cwd(), 'pnpm-workspace.yaml');
    const file = readFileSync(path, 'utf8');
    const parsed = parse(file) || {};

    // 确保 catalog 和 catalogs 是对象
    return {
      catalog:
        parsed.catalog && typeof parsed.catalog === 'object'
          ? parsed.catalog
          : {},
      catalogs:
        parsed.catalogs && typeof parsed.catalogs === 'object'
          ? parsed.catalogs
          : {},
    };
  } catch (error) {
    console.error('Failed to read pnpm-workspace.yaml:', error);

    return { catalog: {}, catalogs: {} };
  }
}

/**
 * 读取 package.json 文件
 * @param {string} url - 文件 URL 或 '~' 表示当前目录
 * @returns {object} package.json 内容
 */
function readPackageJson(url) {
  try {
    const path =
      url === '~'
        ? join(process.cwd(), 'package.json')
        : join(fileURLToPath(url), '../package.json');

    return JSON.parse(readFileSync(path, 'utf8'));
  } catch (error) {
    console.error('Failed to read package.json:', error);

    return {};
  }
}

/**
 * 默认的自定义类型配置
 * @type {object}
 */
const DEFAULT_CUSTOM_TYPES = {
  engines: {
    path: 'engines',
    strategy: 'versionsByName',
  },
  packageManager: {
    path: 'packageManager',
    strategy: 'name@version',
  },
  optional: {
    path: 'optionalDependencies',
    strategy: 'versionsByName',
  },
};

/**
 * 默认的 semver 组配置
 * @type {object[]}
 */
const DEFAULT_SEMVER_GROUPS = [
  {
    dependencies: [
      '@playwright/*',
      '@sentry/*',
      '@types/react-dom',
      '@types/react',
      'playwright-*',
      'react-dom',
      'react',
      'typescript',
      'vue-router',
      'vue',
      '@vue/*',
    ].sort(),
    range: '~',
  },
  {
    dependencyTypes: ['local', 'optional', 'packageManager'],
    range: '',
  },
  {
    range: '^',
  },
];

/**
 * 创建工作区版本组配置
 * @param {object[]} workspaces - 工作区列表
 * @returns {object|undefined} 版本组配置
 */
function createWorkspaceVersionGroup(workspaces) {
  if (workspaces.length === 0) {
    return undefined;
  }

  return {
    dependencies: workspaces.map((item) => item.package.name),
    dependencyTypes: ['!local'],
    label: 'Pin pnpm workspace',
    pinVersion: 'workspace:^',
  };
}

/**
 * 创建目录版本组配置
 * @param {object} packagesMap - 包名到版本的映射对象
 * @param {object} options - 配置选项
 * @param {string} [prefix] - 版本前缀
 * @returns {object|undefined} 版本组配置
 */
function createCatalogVersionGroup(packagesMap, prefix = '') {
  const packageNames =
    packagesMap && typeof packagesMap === 'object'
      ? Object.keys(packagesMap)
      : [];

  if (packageNames.length === 0) {
    return undefined;
  }

  return {
    dependencies: packageNames,
    dependencyTypes: ['dev', 'prod', 'optional', 'peer', 'pnpmOverrides'],
    label: `Pin pnpm catalog ${prefix}`,
    pinVersion: prefix ? `catalog:${prefix}` : 'catalog:',
  };
}

/**
 * 从 catalog 和 catalogs 字段创建所有目录版本组配置
 * @param {object} options - 包含 catalog 和 catalogs 的对象
 * @param {object} [options.catalog={}] - 主目录包版本映射
 * @param {object} [options.catalogs={}] - 目录组映射，键为组名，值为包版本映射
 * @returns {object[]} 版本组配置列表
 */
function createCatalogGroups({ catalog = {}, catalogs = {} }) {
  const groups = [];

  // 处理主 catalog
  const mainGroup = createCatalogVersionGroup(catalog);

  if (mainGroup) {
    groups.push(mainGroup);
  }

  // 处理 catalogs 中的每个组
  for (const [key, packagesMap] of Object.entries(catalogs)) {
    const groupConfig = createCatalogVersionGroup(packagesMap, key);

    if (groupConfig) {
      groups.push(groupConfig);
    }
  }

  return groups;
}

/**
 * 创建 Node.js 引擎版本组配置
 * @param {object} pkg - package.json 内容
 * @returns {object} 版本组配置
 */
function createNodeEngineVersionGroup(pkg = {}) {
  return {
    dependencies: ['node'],
    dependencyTypes: ['engines'],
    label: 'Pin engines.node',
    pinVersion: pkg.engines?.node || '^20.0.0',
  };
}

/**
 * 默认的其他依赖版本组配置
 * @type {object}
 */
const DEFAULT_OTHERS_VERSION_GROUP = {
  dependencyTypes: ['!local'],
  label: 'Pin others',
  preferVersion: 'highestSemver',
};

/**
 * 定义 syncpack 配置
 * @param {string} url - 配置文件 URL
 * @param {object} config - 初始配置
 * @returns {object} 最终配置
 */
export function defineConfig(url, config = {}) {
  // 读取 package.json
  const pkg = readPackageJson(url);
  const isPnpm = pkg.packageManager?.startsWith?.('pnpm');

  // 构建版本组配置
  const versionGroups = [
    // 工作区配置
    isPnpm ? createWorkspaceVersionGroup(findWorkspaces() ?? []) : null,

    // 目录配置
    ...(isPnpm ? createCatalogGroups(readYaml()) : []),

    // 用户配置
    ...(config.versionGroups || []),

    // Node 引擎配置
    createNodeEngineVersionGroup(pkg),

    // 其他依赖配置
    DEFAULT_OTHERS_VERSION_GROUP,
  ].filter(Boolean);

  // 返回最终配置
  return {
    ...config,
    lintFormatting: false,
    customTypes: { ...DEFAULT_CUSTOM_TYPES, ...config.customTypes },
    semverGroups: [...(config.semverGroups || []), ...DEFAULT_SEMVER_GROUPS],
    versionGroups,
  };
}
