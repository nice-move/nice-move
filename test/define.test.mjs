import { readFileSync } from 'node:fs';

import { defineConfig } from '@nice-move/syncpack-config/define.mjs';
import { findWorkspaces } from 'find-workspaces';
import { expect, test, vi } from 'vitest';

// Mock node:fs
vi.mock('node:fs', () => ({
  readFileSync: vi.fn(),
}));

// Mock find-workspaces
vi.mock('find-workspaces', () => ({
  findWorkspaces: vi.fn(),
}));

test('should generate config for pnpm project', () => {
  // Mock package.json
  readFileSync.mockReturnValueOnce(
    JSON.stringify({
      packageManager: 'pnpm@8.0.0',
      engines: {
        node: '^18.0.0',
      },
    }),
  );

  // Mock pnpm-workspace.yaml
  readFileSync.mockReturnValueOnce(`
packages:
  - packages/*
  - modules/*

catalog:
  typescript: ~5.8.3
  vue: ^3.0.0

catalogs:
  overrides:
    eslint: ^9.29.0
    `);

  // Mock workspaces
  findWorkspaces.mockReturnValue([
    { package: { name: '@scope/package-a' } },
    { package: { name: '@scope/package-b' } },
  ]);

  const config = defineConfig('~', {
    versionGroups: [
      {
        dependencies: ['abc'],
        pinVersion: '0.12.23',
      },
    ],
  });

  expect(config).toMatchSnapshot();
});

test('should generate config for non-pnpm project', () => {
  // Mock package.json
  readFileSync.mockReturnValueOnce(
    JSON.stringify({
      packageManager: 'npm@9.0.0',
      engines: {
        node: '^16.0.0',
      },
    }),
  );

  const config = defineConfig('~', {
    versionGroups: [
      {
        dependencies: ['abc'],
        pinVersion: '0.12.23',
      },
    ],
  });

  expect(config).toMatchSnapshot();
});

test('should handle file read errors', () => {
  // Mock package.json read error
  readFileSync.mockImplementationOnce(() => {
    throw new Error('File not found');
  });

  const config = defineConfig('~', {
    versionGroups: [
      {
        dependencies: ['abc'],
        pinVersion: '0.12.23',
      },
    ],
  });

  expect(config).toMatchSnapshot();
});

test('should merge custom config', () => {
  // Mock package.json
  readFileSync.mockReturnValueOnce(
    JSON.stringify({
      packageManager: 'pnpm@8.0.0',
    }),
  );

  // Mock pnpm-workspace.yaml
  readFileSync.mockReturnValueOnce(`
packages:
  - packages/*
    `);

  const config = defineConfig('~', {
    customTypes: {
      custom: {
        path: 'customPath',
        strategy: 'customStrategy',
      },
    },
    semverGroups: [
      {
        dependencies: ['custom-dep'],
        range: '~',
      },
    ],
    versionGroups: [
      {
        dependencies: ['custom-version'],
        label: 'Custom Version',
      },
    ],
  });

  expect(config).toMatchSnapshot();
});

test('should handle empty workspace and catalog', () => {
  // Mock package.json
  readFileSync.mockReturnValueOnce(
    JSON.stringify({
      packageManager: 'pnpm@8.0.0',
    }),
  );

  // Mock empty pnpm-workspace.yaml
  readFileSync.mockReturnValueOnce('{}');

  // Mock empty workspaces
  findWorkspaces.mockReturnValue([]);

  const config = defineConfig('~', {
    versionGroups: [
      {
        dependencies: ['abc'],
        pinVersion: '0.12.23',
      },
    ],
  });

  expect(config).toMatchSnapshot();
});
