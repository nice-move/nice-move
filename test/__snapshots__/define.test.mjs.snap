// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should generate config for non-pnpm project 1`] = `
{
  "customTypes": {
    "engines": {
      "path": "engines",
      "strategy": "versionsByName",
    },
    "optional": {
      "path": "optionalDependencies",
      "strategy": "versionsByName",
    },
    "packageManager": {
      "path": "packageManager",
      "strategy": "name@version",
    },
  },
  "lintFormatting": false,
  "semverGroups": [
    {
      "dependencies": [
        "@playwright/*",
        "@sentry/*",
        "@types/react",
        "@types/react-dom",
        "@vue/*",
        "playwright-*",
        "react",
        "react-dom",
        "typescript",
        "vue",
        "vue-router",
      ],
      "range": "~",
    },
    {
      "dependencyTypes": [
        "local",
        "optional",
        "packageManager",
      ],
      "range": "",
    },
    {
      "range": "^",
    },
  ],
  "versionGroups": [
    {
      "dependencies": [
        "abc",
      ],
      "pinVersion": "0.12.23",
    },
    {
      "dependencies": [
        "node",
      ],
      "dependencyTypes": [
        "engines",
      ],
      "label": "Pin engines.node",
      "pinVersion": "^16.0.0",
    },
    {
      "dependencyTypes": [
        "!local",
      ],
      "label": "Pin others",
      "preferVersion": "highestSemver",
    },
  ],
}
`;

exports[`should generate config for pnpm project 1`] = `
{
  "customTypes": {
    "engines": {
      "path": "engines",
      "strategy": "versionsByName",
    },
    "optional": {
      "path": "optionalDependencies",
      "strategy": "versionsByName",
    },
    "packageManager": {
      "path": "packageManager",
      "strategy": "name@version",
    },
  },
  "lintFormatting": false,
  "semverGroups": [
    {
      "dependencies": [
        "@playwright/*",
        "@sentry/*",
        "@types/react",
        "@types/react-dom",
        "@vue/*",
        "playwright-*",
        "react",
        "react-dom",
        "typescript",
        "vue",
        "vue-router",
      ],
      "range": "~",
    },
    {
      "dependencyTypes": [
        "local",
        "optional",
        "packageManager",
      ],
      "range": "",
    },
    {
      "range": "^",
    },
  ],
  "versionGroups": [
    {
      "dependencies": [
        "@scope/package-a",
        "@scope/package-b",
      ],
      "dependencyTypes": [
        "!local",
      ],
      "label": "Pin pnpm workspace",
      "pinVersion": "workspace:^",
    },
    {
      "dependencies": [
        "typescript",
        "vue",
      ],
      "dependencyTypes": [
        "dev",
        "prod",
        "optional",
        "peer",
        "pnpmOverrides",
      ],
      "label": "Pin pnpm catalog ",
      "pinVersion": "catalog:",
    },
    {
      "dependencies": [
        "eslint",
      ],
      "dependencyTypes": [
        "dev",
        "prod",
        "optional",
        "peer",
        "pnpmOverrides",
      ],
      "label": "Pin pnpm catalog overrides",
      "pinVersion": "catalog:overrides",
    },
    {
      "dependencies": [
        "abc",
      ],
      "pinVersion": "0.12.23",
    },
    {
      "dependencies": [
        "node",
      ],
      "dependencyTypes": [
        "engines",
      ],
      "label": "Pin engines.node",
      "pinVersion": "^18.0.0",
    },
    {
      "dependencyTypes": [
        "!local",
      ],
      "label": "Pin others",
      "preferVersion": "highestSemver",
    },
  ],
}
`;

exports[`should handle empty workspace and catalog 1`] = `
{
  "customTypes": {
    "engines": {
      "path": "engines",
      "strategy": "versionsByName",
    },
    "optional": {
      "path": "optionalDependencies",
      "strategy": "versionsByName",
    },
    "packageManager": {
      "path": "packageManager",
      "strategy": "name@version",
    },
  },
  "lintFormatting": false,
  "semverGroups": [
    {
      "dependencies": [
        "@playwright/*",
        "@sentry/*",
        "@types/react",
        "@types/react-dom",
        "@vue/*",
        "playwright-*",
        "react",
        "react-dom",
        "typescript",
        "vue",
        "vue-router",
      ],
      "range": "~",
    },
    {
      "dependencyTypes": [
        "local",
        "optional",
        "packageManager",
      ],
      "range": "",
    },
    {
      "range": "^",
    },
  ],
  "versionGroups": [
    {
      "dependencies": [
        "abc",
      ],
      "pinVersion": "0.12.23",
    },
    {
      "dependencies": [
        "node",
      ],
      "dependencyTypes": [
        "engines",
      ],
      "label": "Pin engines.node",
      "pinVersion": "^20.0.0",
    },
    {
      "dependencyTypes": [
        "!local",
      ],
      "label": "Pin others",
      "preferVersion": "highestSemver",
    },
  ],
}
`;

exports[`should handle file read errors 1`] = `
{
  "customTypes": {
    "engines": {
      "path": "engines",
      "strategy": "versionsByName",
    },
    "optional": {
      "path": "optionalDependencies",
      "strategy": "versionsByName",
    },
    "packageManager": {
      "path": "packageManager",
      "strategy": "name@version",
    },
  },
  "lintFormatting": false,
  "semverGroups": [
    {
      "dependencies": [
        "@playwright/*",
        "@sentry/*",
        "@types/react",
        "@types/react-dom",
        "@vue/*",
        "playwright-*",
        "react",
        "react-dom",
        "typescript",
        "vue",
        "vue-router",
      ],
      "range": "~",
    },
    {
      "dependencyTypes": [
        "local",
        "optional",
        "packageManager",
      ],
      "range": "",
    },
    {
      "range": "^",
    },
  ],
  "versionGroups": [
    {
      "dependencies": [
        "abc",
      ],
      "pinVersion": "0.12.23",
    },
    {
      "dependencies": [
        "node",
      ],
      "dependencyTypes": [
        "engines",
      ],
      "label": "Pin engines.node",
      "pinVersion": "^20.0.0",
    },
    {
      "dependencyTypes": [
        "!local",
      ],
      "label": "Pin others",
      "preferVersion": "highestSemver",
    },
  ],
}
`;

exports[`should merge custom config 1`] = `
{
  "customTypes": {
    "custom": {
      "path": "customPath",
      "strategy": "customStrategy",
    },
    "engines": {
      "path": "engines",
      "strategy": "versionsByName",
    },
    "optional": {
      "path": "optionalDependencies",
      "strategy": "versionsByName",
    },
    "packageManager": {
      "path": "packageManager",
      "strategy": "name@version",
    },
  },
  "lintFormatting": false,
  "semverGroups": [
    {
      "dependencies": [
        "custom-dep",
      ],
      "range": "~",
    },
    {
      "dependencies": [
        "@playwright/*",
        "@sentry/*",
        "@types/react",
        "@types/react-dom",
        "@vue/*",
        "playwright-*",
        "react",
        "react-dom",
        "typescript",
        "vue",
        "vue-router",
      ],
      "range": "~",
    },
    {
      "dependencyTypes": [
        "local",
        "optional",
        "packageManager",
      ],
      "range": "",
    },
    {
      "range": "^",
    },
  ],
  "versionGroups": [
    {
      "dependencies": [
        "@scope/package-a",
        "@scope/package-b",
      ],
      "dependencyTypes": [
        "!local",
      ],
      "label": "Pin pnpm workspace",
      "pinVersion": "workspace:^",
    },
    {
      "dependencies": [
        "custom-version",
      ],
      "label": "Custom Version",
    },
    {
      "dependencies": [
        "node",
      ],
      "dependencyTypes": [
        "engines",
      ],
      "label": "Pin engines.node",
      "pinVersion": "^20.0.0",
    },
    {
      "dependencyTypes": [
        "!local",
      ],
      "label": "Pin others",
      "preferVersion": "highestSemver",
    },
  ],
}
`;
